linera-views/src/scylla_db.rs

```diff
@@ -164,34 +164,81 @@ impl ScyllaDbClient {
+                    let entry = entry.into_mut();
+                    entry.push(i_key);
+                }
+                Entry::Vacant(entry) => {
+                    entry.insert(vec![i_key]);
+                    unique_keys.push(key);
+                }
+            }
+        }
+        let num_unique_keys = map.len();
         let mut group_query = "?".to_string();
-        group_query.push_str(&",?".repeat(num_keys - 1));
+        group_query.push_str(&",?".repeat(num_unique_keys - 1));
         let query = format!(
             "SELECT k,v FROM kv.{} WHERE dummy = 0 AND k IN ({}) ALLOW FILTERING",
             self.namespace, group_query
         );
         let read_multi_value = Query::new(query);
-        let mut rows = session.query_iter(read_multi_value, &keys).await?;
+        let mut rows = session.query_iter(read_multi_value, &unique_keys).await?;
         let mut values = vec![None; num_keys];
+        while let Some(row) = rows.next().await {
+            let value = row?.into_typed::<(Vec<u8>, Vec<u8>)>()?;
+            let key = value.0;
+            for i_key in map.get(&key).unwrap().clone() {
+                let value = Some(value.1.clone());
+                *values.get_mut(i_key).expect("an entry in values") = value;
+            }
+        }
+        Ok(values)
+    }
+
+    async fn contains_keys_internal(
+        &self,
+        keys: Vec<Vec<u8>>,
+    ) -> Result<Vec<bool>, ScyllaDbStoreError> {
+        let num_keys = keys.len();
+        if num_keys == 0 {
+            return Ok(Vec::new());
+        }
+        let session = &self.session;
         let mut map = HashMap::<Vec<u8>, Vec<usize>>::new();
+        let mut unique_keys = Vec::new();
         for (i_key, key) in keys.into_iter().enumerate() {
             ensure!(key.len() <= MAX_KEY_SIZE, ScyllaDbStoreError::KeyTooLong);
-            match map.entry(key) {
+            match map.entry(key.clone()) {
                 Entry::Occupied(entry) => {
                     let entry = entry.into_mut();
                     entry.push(i_key);
                 }
                 Entry::Vacant(entry) => {
                     entry.insert(vec![i_key]);
+                    unique_keys.push(key);
                 }
             }
         }
+        let num_unique_keys = map.len();
+        let mut group_query = "?".to_string();
+        group_query.push_str(&",?".repeat(num_unique_keys - 1));
+        let query = format!(
+            "SELECT k FROM kv.{} WHERE dummy = 0 AND k IN ({}) ALLOW FILTERING",
+            self.namespace, group_query
+        );
+        let contains_keys = Query::new(query);
+        let mut rows = session.query_iter(contains_keys, &unique_keys).await?;
+        let mut values = vec![false; num_keys];
         while let Some(row) = rows.next().await {
-            let value = row?.into_typed::<(Vec<u8>, Vec<u8>)>()?;
+            let value = row?.into_typed::<(Vec<u8>,)>()?;
             let key = value.0;
```

Comment:
```suggestion
            let (key,) = row?.into_typed::<(Vec<u8>,)>()?;
```

Reply:
I did not know that syntax. Thanks. It has been applied at other places in the code.